src/Bicep.Core/Emit/ExpressionConverter.cs

```diff
@@ -584,13 +605,10 @@ private LanguageExpression ConvertVariableAccess(VariableAccessSyntax variableAc
                     }
                     return CreateFunction("variables", new JTokenExpression(name));
 
-                case ResourceSymbol resourceSymbol when context.SemanticModel.ResourceMetadata.TryLookup(variableAccessSyntax) is {} resource:
+                case ResourceSymbol when context.SemanticModel.ResourceMetadata.TryLookup(variableAccessSyntax) is { } resource:
                     return GetReferenceExpression(resource, true);
 
                 case ModuleSymbol moduleSymbol:
-                    // referencing a module directly should be blocked at an earlier stage - there is nothing great we can codegen here.
-                    Debug.Fail("Found direct module variable access");
```

Comment:
I've created https://github.com/Azure/bicep/issues/4320 just to make sure we keep track of the original problem.